designs/2022-supress-ignored-file-warnings/README.md

```diff
@@ -0,0 +1,157 @@
+- Repo: eslint/eslint
+- Start Date: 2022-05-13
+- RFC PR:
+- Authors: Domantas Petrauskas
+
+# CLI option to supress ignored file warnings
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+When ignored files are explicitly linted, ESLint shows a warning. This causes problems when using tools which pass the file list to ESLint automatically together with `--max-warnings 0` option.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+While this is warning is reasonable in cases where filenames are passed to ESLint manually, it causes issues when the process is automated with tools like [lint-staged](https://github.com/okonet/lint-staged) and [pre-commit](https://github.com/pre-commit/pre-commit). These tools pass staged filenames to any command, not just ESLint. Therefore, they are not aware of the .eslintignore. Linting before commit is commonly set up with `--max-warnings 0`, therefore ESLint will exit with an error in such case.
+
+For example, with
+
+```
+// .eslintignore
+foo.js
+```
+
+Running
+
+```
+eslint foo.js
+```
+
+Will result in
+
+```
+warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override
+```
+
+It is possible to filter out ignored files using ESLint `CLIEngine`. This was the main reasoning to reject the CLI flag idea during [TSC meeting in 2018](https://gitter.im/eslint/tsc-meetings/archives/2018/08/02). The commitee was only considering the use case of integrations. However, this problem affects end users as well. Lint-staged FAQ includes a section [how can I ignore files from .eslintignore](https://github.com/okonet/lint-staged#how-can-i-ignore-files-from-eslintignore), which suggests adding additional config file and using `CLIEngine.isPathIgnored` to filter out ignored files. Having a CLI flag for this would improve the end-user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+ESLint CLI should have an optional flag to disable the ignored file warning. The flag should be boolean. In order to stay consistent with current CLI flags, it should be called `--no-warning-on-ignored-files`. If it is present, ESLint should supress the `File ignored because of a matching ignore pattern.` warning. The flag should replicate `eslint.lintText`[warnIgnored option](https://eslint.org/docs/developer-guide/nodejs-api#-eslintlinttextcode-options) when it is set to `false`.
```

Comment:
how is it working with the `--no-error-on-unmatched-pattern`?

Reply:
Good question. As far as I see they should work completely separately, but I'll check the code to see if that's really the case

Reply:
After looking at the code, I can confirm `--no-error-on-unmatched-pattern` should work independently from `--no-warning-on-ignored-files`.

Let's say I have an `index.js` file.

---
```
eslint index.js --ignore-pattern index.js
```
will result in
```
warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override
```
---
```
eslint index.js --ignore-pattern index.js --no-warning-on-ignored-files
```
will result in no errors/warnings, the file will not be linted because it is ignored.

---
```
eslint nonexistingfile.js --ignore-pattern index.js
```
will result in error:
```
Oops! Something went wrong! :(

ESLint: 8.12.0

No files matching the pattern "asdfasdf.js" were found.
Please check for typing mistakes in the pattern.
```
--- 
```
eslint nonexistingfile.js --ignore-pattern index.js --no-error-on-unmatched-pattern
```
will result in no errors/warnings _because the unmatched pattern error is suppressed_

---
```
eslint nonexistingfile.js --ignore-pattern nonexistingfile.js --no-error-on-unmatched-pattern --no-warning-on-ignored-files
```
will result in no errors/warnings _because the unmatched pattern error is suppressed_. `--no-warning-on-ignored-files` will have no effect, because eslint exits on unmatched pattern before processing ignored files.

Reply:
is it possible to merge the two options into one?

```bash
--on-unmatched-pattern=error/warn/ignore

--on-unmatched-pattern=error #default option
--on-unmatched-pattern=warn #same as the --no-error-on-unmatched-pattern
--on-unmatched-pattern=ignore #same as the --no-warning-on-ignored-files
```